File: macos/Sources/ElevateCore/Storage/AppStateStore.swift
import Foundation

public actor AppStateStore {
    private let fileURL: URL
    /// Generation of the newest state written; out-of-order saves are dropped.
    private var lastAppliedGeneration: UInt64 = 0
    private var nextGeneration: UInt64 = 0

    public static var defaultDirectory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let current = base.appendingPathComponent("Elevate", isDirectory: true)
        let legacy = base.appendingPathComponent("PimTray", isDirectory: true)
        // One-time move of state saved under the app's former name.
        if !FileManager.default.fileExists(atPath: current.path), FileManager.default.fileExists(atPath: legacy.path) {
            try? FileManager.default.moveItem(at: legacy, to: current)
        }
        return current
    }

    public init(directory: URL = AppStateStore.defaultDirectory) {
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        fileURL = directory.appendingPathComponent("state.json")
    }

    public func load() throws -> AppState {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return AppState() }
        return try GraphJSON.decoder.decode(AppState.self, from: Data(contentsOf: fileURL))
    }

    /// Moves an unreadable `state.json` aside so the next save cannot silently destroy it.
    /// Returns the backup URL, or nil when there was no file to move.
    public func quarantineCorruptFile() throws -> URL? {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return nil }
        let backup = fileURL.deletingLastPathComponent().appendingPathComponent("state.json.bak")
        try? FileManager.default.removeItem(at: backup)
        try FileManager.default.moveItem(at: fileURL, to: backup)
        return backup
    }

    public func save(_ state: AppState) throws {
        nextGeneration += 1
        try save(state, generation: nextGeneration)
    }

    /// Writes `state` unless a newer generation has already been written.
    public func save(_ state: AppState, generation: UInt64) throws {
        guard generation >= lastAppliedGeneration else { return }
        let encoder = GraphJSON.encoder
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(state).write(to: fileURL, options: .atomic)
        lastAppliedGeneration = generation
        nextGeneration = max(nextGeneration, generation)
    }
}

File: windows/src/Elevate.Core/Storage/AppStateStore.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Elevate.Core.Storage;

/// <summary>
/// Reads and writes <c>state.json</c>. Writes are atomic (temp file plus move) and carry a
/// generation, so a save computed from older state cannot overwrite a newer one.
/// </summary>
public sealed class AppStateStore
{
    private readonly Lock _gate = new();
    private readonly string _filePath;

    /// <summary>Generation of the newest state written; out-of-order saves are dropped.</summary>
    private ulong _lastAppliedGeneration;
    private ulong _nextGeneration;

    public AppStateStore(string? directory = null)
    {
        Directory = directory ?? DefaultDirectory;
        System.IO.Directory.CreateDirectory(Directory);
        _filePath = Path.Combine(Directory, "state.json");
    }

    /// <summary><c>%LOCALAPPDATA%\Elevate</c> on Windows; the platform equivalent elsewhere.</summary>
    public static string DefaultDirectory =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Elevate");

    public string Directory { get; }

    public string FilePath => _filePath;

    /// <summary>The saved state, or an empty one when nothing has been written yet.</summary>
    public AppState Load()
    {
        lock (_gate)
        {
            if (!File.Exists(_filePath))
            {
                return new AppState();
            }

            var bytes = File.ReadAllBytes(_filePath);
            return JsonSerializer.Deserialize<AppState>(bytes, Json.Options) ?? new AppState();
        }
    }

    /// <summary>
    /// Moves an unreadable <c>state.json</c> aside so the next save cannot silently destroy it.
    /// Returns the backup path, or null when there was no file to move.
    /// </summary>
    public string? QuarantineCorruptFile()
    {
        lock (_gate)
        {
            if (!File.Exists(_filePath))
            {
                return null;
            }

            var backup = Path.Combine(Directory, "state.json.bak");
            File.Move(_filePath, backup, overwrite: true);
            return backup;
        }
    }

    public void Save(AppState state)
    {
        lock (_gate)
        {
            _nextGeneration += 1;
            SaveLocked(state, _nextGeneration);
        }
    }

    /// <summary>Writes <paramref name="state"/> unless a newer generation has already been written.</summary>
    public void Save(AppState state, ulong generation)
    {
        lock (_gate)
        {
            SaveLocked(state, generation);
        }
    }

    private void SaveLocked(AppState state, ulong generation)
    {
        if (generation < _lastAppliedGeneration)
        {
            return;
        }

        var temp = _filePath + ".tmp";
        File.WriteAllBytes(temp, Encode(state));
        File.Move(temp, _filePath, overwrite: true);
        _lastAppliedGeneration = generation;
        _nextGeneration = Math.Max(_nextGeneration, generation);
    }

    /// <summary>
    /// Pretty-printed with sorted keys, matching the macOS encoder's
    /// <c>[.prettyPrinted, .sortedKeys]</c> output so the same file is produced on both platforms.
    /// </summary>
    internal static byte[] Encode(AppState state)
    {
        var node = JsonSerializer.SerializeToNode(state, Json.Options);
        var sorted = Sorted(node);
        var text = sorted?.ToJsonString(new JsonSerializerOptions(Json.Options) { WriteIndented = true }) ?? "{}";
        return Encoding.UTF8.GetBytes(text);
    }

    private static JsonNode? Sorted(JsonNode? node)
    {
        switch (node)
        {
            case JsonObject o:
            {
                var result = new JsonObject();
                foreach (var (key, value) in o.ToList().OrderBy(p => p.Key, StringComparer.Ordinal))
                {
                    o.Remove(key);
                    result[key] = Sorted(value);
                }

                return result;
            }

            case JsonArray a:
            {
                var items = a.ToList();
                var result = new JsonArray();
                foreach (var item in items)
                {
                    a.Remove(item);
                    result.Add(Sorted(item));
                }

                return result;
            }

            default:
                return node;
        }
    }
}
